docs/src/references/agent/operation-workflow.md

```diff
@@ -555,62 +554,125 @@ action = "cleanup"
+
+The workflows of these builtin operations can be customized.
+
+For each, there are *two* workflows: an internal workflow and a customized version of the former.
+- The `builtin:<operation-name>` operation workflow describes the builtin behavior as provided by __tedge-agent__.
+- The `<operation-name>` operation workflow rules the actual behavior of __tedge-agent__,
+  possibly invoking `builtin:<operation-name>` behind the scene.
+- When a user or a mapper triggers a command over MQTT for that `<operation-name>`,
+  by publishing an init message on `te/+/+/+/+/cmd/builtin:<operation-name>/+`,
+  __tedge_agent__ uses the customized version of the workflow for that operation.
+  - From MQTT, there is no way to trigger directly the builtin version of an operation,
+    i.e. there is no `te/+/+/+/+/cmd/builtin:<operation-name>/+` command topic.
+  - The builtin versions are only internal triggered by the agent from a customized workflow. 
+- On a fresh install of %%te%%, the `<operation-name>` is simply a copy of `builtin:<operation-name>`.
+  - This copy is not materialized on the file system, but created in memory by the agent when no customized version is found.
+- A customized version for an `<operation-name>` is provided as a TOML workflow definition file in `/etc/tedge/operations`
+  for that operation, as for any user provided workflow.
+  - By convention, this file is named after the operation name, as in `/etc/tedge/operations/software_update.toml`.
+    However, this is not mandatory: the operation is determined by the `operation` property, e.g. `operation = "software_update"`.
+  - If this definition is valid, then it will be used by the agent as the `<operation-name>` workflow.
+- The customized version of an `<operation-name>` can invoke its builtin version
+  with a step triggering `operation = "builtin:<operation-name>"`.
+  - A customized workflow can also be a complete rework of the feature, ignoring the builtin behavior.
+- A user-provided workflow can invoke any builtin operations as well as the custom versions.
+  - `operation = "<operation-name>"` triggers the custom-version of the operation workflow.
+  - `operation = "builtin:<operation-name>"` triggers the operation ignoring any customization.
+
+In order to customize a builtin operation, the first step is to materialize its definition in `/etc/tedge/operations`.
+For instance, here is the builtin workflow for the `software_update` operation:
+
+```toml title="/etc/tedge/operations/software_update.toml"
+operation = "software_update"            # any builtion operation can be customized
 
 ["init"]
-action = "proceed"                  # open to customization
+action = "proceed"                       # open to customization
 on_success = "scheduled"
 
 [scheduled]
-action = "builtin"                  # delegated to the tedge-agent
-on_success = "executing"
+operation = "builtin:software_update"    # trigger the built-in behavior for software update
```

Comment:
> Seeing this indirect tying of the `builtin:software_update` operation to the `scheduled` state and `await-operation-completion` to the `executing` state, it doesn't look very straightforward, although I understand how the current impl is forcing us to do this. Ignoring the impl side of things, but purely from a workflow user's perspective

With this proposal *doesn't tie* the `builtin:software_update` to one state or another (by contrast to what is done by the deprecated `builtin` action).

>  I'd have expected the `builtin:software_update` operation to be triggered from/after the `executing` state (since that's the logical start point of an operation execution in my mind):

As long as the workflow contains an `executing` state to tell c8y that the operations has been accepted,
you can trigger the builtin operation on any state. 

Notably, what you propose should work. => I will add a system test for that.

```toml
[scheduled]
action = "proceed"
on_success = "executing"

[executing]
operation = "builtin:software_update"    # trigger the built-in behavior for software update
on_exec = "await_operation"

[await_operation]
action = "await-operation-completion"    # awaiting the builtin operation to complete
on_success = "successful"

```

> But, since our bult-in actors do parts of the operation processing in the `scheduled` state like the `tedgeUrl` generation and the rest in the `executing` state, we're forced to define things as proposed. But, for someone who doesn't know the internal impl, it definitely wouldn't be that straightforward.

No, what you describe is the behavior of the deprecated `builtin` action. And the [defunct proposal](https://github.com/thin-edge/thin-edge.io/pull/3070) was awkwardly trying to hide this with two different actions (one to trigger a builtin `schedule` and the other a builtin `execute`).

"builtin:software_update" triggers the *whole* workflow, not just one step, and await for the "successful" or "failed" state.

> Something as follows would at least make this internal contract explicit:

What you describe with  `operation_action = "builtin:software_update:scheduled"` and `operation_action = "builtin:software_update:executing"` is the implementation of the  [defunct proposal](https://github.com/thin-edge/thin-edge.io/pull/3070)

> But this is too verbose and possibly exposing internal impl details too much, when the user doesn't care about the state separation. They just want the `builtin:software_update` to be executed end-to-end from any state of their choice.

Precisely, this is why we move to this proposal, where the `builtin:software_update` is executed end-to-end.

To be honest, the current implement is jumping straight to the schedule state, as I know that all the builtin operation are doing something before.

> For me, if the user is free to call the `builtin_<operation>` from any state(from/after the `executing` state, of course), that's the ideal outcome.

The user is free to do so. As already said, I will add a system test for that.

Reply:
> Seeing this indirect tying of the builtin:software_update operation to the scheduled state and await-operation-completion to the executing state, it doesn't look very straightforward, although I understand how the current impl is forcing us to do this.

Sorry, it was a false alarm. I messed up something during my initial testing, where the `builtin` operation was only working when specified in the `scheduled` state and failing with the following error when I moved it to the `executing` state:

```
Sep 06 11:53:28 3e42bac35040 tedge-agent[611]: 2024-09-06T11:53:28.475093579Z  WARN tedge_agent::operation_file_cache: Received config update, but payload is malformed: EOF while parsing a value at line 1 column 0
```

which made me think that we still have that limitation with states. But, when I tested everything from scratch, post-your-response, everything seemed to be working fine, even when configured in the `executing` state.

Reply:
> As long as the workflow contains an executing state to tell c8y that the operations has been accepted,
you can trigger the builtin operation on any state.

> Notably, what you propose should work. => I will add a system test for that.

Here is the test: https://github.com/thin-edge/thin-edge.io/pull/3105/commits/1a8de58129b10f2f7e1a78626c5afc81cf081ce4